cpp/dolfinx/fem/interpolate.h

```diff
@@ -12,20 +12,188 @@
+{
+  assert(src_ranks.size() == send_values.shape(0));
+  const std::size_t value_size = send_values.shape(1);
+  xt::xtensor<T, 2> values({dest_ranks.size(), value_size});
+
+  // Create neighborhood communicator from send back
+  // values to requesting processes
+  // NOTE: source rank is already sorted
+  std::vector<std::int32_t> out_ranks(src_ranks);
+  out_ranks.erase(std::unique(out_ranks.begin(), out_ranks.end()),
+                  out_ranks.end());
+  out_ranks.reserve(out_ranks.size() + 1);
+
+  // Strip dest ranks for all -1 entries
+  std::vector<std::int32_t> in_ranks;
+  in_ranks.reserve(dest_ranks.size());
+  std::vector<std::int32_t> rank_mapping;
+  rank_mapping.reserve(dest_ranks.size());
+  for (std::size_t i = 0; i < dest_ranks.size(); ++i)
+  {
+    if (const std::int32_t rank = dest_ranks[i]; rank >= 0)
+    {
+      rank_mapping.push_back(i);
+      in_ranks.push_back(rank);
+    }
+  }
+
+  // Create unique set of sorted in ranks
+  std::sort(in_ranks.begin(), in_ranks.end());
+  in_ranks.erase(std::unique(in_ranks.begin(), in_ranks.end()), in_ranks.end());
+  in_ranks.reserve(in_ranks.size() + 1);
+
+  // Create communicator from processes with dof values to the processes
+  // owning the dofs
+  MPI_Comm reverse_comm;
+  MPI_Dist_graph_create_adjacent(
+      comm, in_ranks.size(), in_ranks.data(), MPI_UNWEIGHTED, out_ranks.size(),
+      out_ranks.data(), MPI_UNWEIGHTED, MPI_INFO_NULL, false, &reverse_comm);
+
+  // Compute map from incoming value to local point index
+  std::vector<std::int32_t> unpack_map;
+  std::vector<std::int32_t> recv_sizes(in_ranks.size());
+  recv_sizes.reserve(1);
+  std::vector<std::int32_t> recv_offsets(in_ranks.size() + 1, 0);
+  {
+    std::map<std::int32_t, std::int32_t> rank_to_neighbor;
+    for (std::size_t i = 0; i < in_ranks.size(); i++)
+      rank_to_neighbor[in_ranks[i]] = i;
+    for (std::size_t i = 0; i < rank_mapping.size(); i++)
+    {
+      const int inc_rank = dest_ranks[rank_mapping[i]];
+      const int neighbor = rank_to_neighbor[inc_rank];
+      recv_sizes[neighbor] += value_size;
+    }
+    // Compute receiving offsets
+    std::vector<std::int32_t> recv_counter(recv_sizes.size(), 0);
+    std::partial_sum(recv_sizes.begin(), recv_sizes.end(),
+                     std::next(recv_offsets.begin(), 1));
+    unpack_map.resize(recv_offsets.back() / value_size);
+    for (std::size_t i = 0; i < rank_mapping.size(); i++)
+    {
+      const int inc_rank = dest_ranks[rank_mapping[i]];
+      const int neighbor = rank_to_neighbor[inc_rank];
+      int pos = recv_offsets[neighbor] + recv_counter[neighbor];
+      unpack_map[pos / value_size] = rank_mapping[i];
+      recv_counter[neighbor] += value_size;
+    }
+  }
+  // Compute map from global mpi rank to neigbor rank for outgoing data
+  std::vector<std::int32_t> send_sizes(out_ranks.size());
+  send_sizes.reserve(1);
+  {
+    std::map<std::int32_t, std::int32_t> rank_to_neighbor;
```

Comment:
Best to avoid `std::maps`. Build vector of pairs, then sort and then use binary search.